boxtree/tree.py

```diff
@@ -111,9 +112,111 @@ class box_flags_enum(Enum):  # noqa
 # }}}
 
 
+# {{{ tree of boxes
+
+class TreeOfBoxes:
+    """A quad/octree tree of abstract boxes. Lightweight trees handled with
+    :mod:`numpy`, intended for mesh adaptivity.
+
+    .. attribute:: dimensions
+
+    .. attribute:: nlevels
+
+    .. attribute:: nboxes
+
+    .. attribute:: root_extent
+
+        (Scalar) extent of the root box.
+
+    .. attribute:: box_centers
+
+        dim-by-nboxes :mod:`numpy` array of the centers of the boxes.
+
+    .. attribute:: box_parent_ids
+
+        :mod:`numpy` vector of parent box ids.
+
+    .. attribute:: box_child_ids
+
+        (2**dim)-by-nboxes :mod:`numpy` array of children box ids.
+
+    .. attribute:: box_levels
+
+        :mod:`numpy` vector of box levels.
+
+    .. automethod:: __init__
+
+    .. automethod:: copy
+
+    .. automethod:: get_leaf_flags
+
+    .. automethod:: leaf_boxes
+    """
+    def __init__(
+            self, box_centers, root_extent,
+            box_parent_ids, box_child_ids, box_levels):
+        self.box_centers = box_centers
+        self.root_extent = root_extent
+        self.box_parent_ids = box_parent_ids
+        self.box_child_ids = box_child_ids
+        self.box_levels = box_levels
+
+        lows = self.box_centers[:, 0] - 0.5*self.root_extent
+        highs = lows + self.root_extent
+        self.bounding_box = [lows, highs]
+
+        self.dim = self.box_centers.shape[0]
+
+    def copy(self):
+        return TreeOfBoxes(
+                box_centers=self.box_centers.copy(),
+                root_extent=self.root_extent.copy(),
+                box_parent_ids=self.box_parent_ids.copy(),
+                box_child_ids=self.box_child_ids.copy(),
+                box_levels=self.box_levels.copy())
+
+    # {{{ dummy interface for TreePlotter
+
+    @property
+    def dimensions(self):
+        return self.dim
+
+    def get_box_size(self, ibox):
+        lev = self.box_levels[ibox]
+        box_size = self.root_extent * 0.5**lev
+        return box_size
+
+    def get_box_extent(self, ibox):
+        box_size = self.get_box_size(ibox)
+        extent_low = self.box_centers[:, ibox] - 0.5*box_size
+        extent_high = extent_low + box_size
+        return extent_low, extent_high
+
+    # }}} End dummy interface for TreePlotter
+
+    @property
+    def nboxes(self):
+        return self.box_centers.shape[1]
+
+    @property
+    def nlevels(self):
+        return max(self.box_levels) + 1  # level starts from 0
+
+    def get_leaf_flags(self):
+        # box_id -> whether the box is leaf
+        return np.all(self.box_child_ids == 0, axis=0)
+
+    def leaf_boxes(self):
+        boxes = np.arange(self.nboxes)
+        return boxes[self.get_leaf_flags()]
+
+
+# }}}
+
+
 # {{{ tree data structure
 
-class Tree(DeviceDataRecord):
+class Tree(DeviceDataRecord, TreeOfBoxes):
```

Comment:
As it is, won't `Tree` inherit a bunch of methods that don't necessarily work on it?

Reply:
Yes, `copy`, `get_leaf_flags`, `leaf_boxes` will be unused. One inelegant solution to this would be to inject those methods in `TreeOfBoxes.__post_init__`. What do you think?

Reply:
I don't find that appealing. I think it's quite possible to make this work properly.

Reply:
Remove the guards. This'll become more OK as we move towards an arraycontext-only world.